unic/utils/src/char_property.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright 2017 The UNIC Project Developers.
+//
+// See the COPYRIGHT file at the top-level directory of this distribution.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+
+//! Taxonomy and Contracts for Character Property types.
+//!
+//! See also the list of types of character properties defined in the UCD:
+//! <http://unicode.org/reports/tr44/#About_Property_Table>, in [Unicode® Standard Annex #44 —
+//! Unicode Character Database](http://unicode.org/reports/tr44/#About_Property_Table)
+
+
+use std::fmt::{Debug, Display};
+use std::hash::Hash;
+
+
+// == Completeness and Accessors ==
+
+/// Marker for character property types.
+pub trait CharProperty<T> {}
+
+// Make char property type T also a parameter to this trait, otherwise there will be "conflicting
+// implementation" error when marking traits below as `CharProperty`.
+impl<T> CharProperty<T> for T {}
```

Comment:
Seems like what you want is `CompleteCharProperty` and `OptionCharProperty` to be mutually exclusive...

Making `OptionCharProperty` a parent of `CompleteCharProperty` and using `OptionCharProperty` would solve the problem (and in any case I suspect `of` will be intrinsic and used as such), but it seems difficult to use this pattern like this.